ColumnMap with O(1) lookup

Enmk opened the pull request:
* Building index with a 1 hashtable per row, but all hashtables share the memory arena, hence are easier on memory allocations.
* Specializations of the ColumnMapIndex for different types of keys.
* Statefull allocators for HashMap and StringHashMap.
* Interface of ColumnString and ColumnFixedString more consistent with ColumnVector and ColumnDecimal.
* With unit tests

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
O(1) lookup for ColumnMap operator []
...


Detailed description / Documentation draft:

...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.


Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

Enmk committed:
Use index to speed up lookup in ColunMap

* Building index with a 1 hashtable per row, but all hashtables share the memory arena, hence are easier on memory allocations.
* Index is buit only only when it is really needed and destroyed on
  column data modification.
* Specializations of the ColumnMapIndex for different types of keys.
* Interface of ColumnString and ColumnFixedString more consistent with ColumnVector and ColumnDecimal.
* With unit tests
* Allowing nullable key type

yuzhichang commented:
I think storing each key to a subcolumn is better.

alexey-milovidov commented:
@yuzhichang Yes, it will be better (for small number of keys) but we don't have support for dynamic subcolumns yet.

Enmk committed:
Minor: fixed style issues

Enmk committed:
Fixed clang-11 reported warnings

Enmk committed:
Attempt to fix implicit index rebuild performance

Index is now rebuilt explicitly and is expected to be always present and
up to date.

Enmk committed:
Updated ArenaWithFreeLists allocator

* Added compile-time options to  support clearing and pre-mapping memory if required
* Fixed implementation of realloc() to take memory size and size class
  into account.

Enmk committed:
Custom compile-time options for Allocator used by Arena

Enmk commented:
Failed to demonstrate any improvement in performance, closing for now
